File: app/src/main/java/com/drawin3d/StrokeProcessing.java
package com.drawin3d;

import java.util.ArrayList;
import java.util.List;

/** Baked authoring operations matching web/src/stroke-processing.js. */
final class StrokeProcessing {
    private static float[] lerp(float[] a,float[] b,double t){float[] p=new float[4];for(int j=0;j<4;j++)p[j]=(float)(a[j]+(b[j]-a[j])*t);return p;}
    private static double distance(float[] a,float[] b){double x=(double)a[0]-b[0],y=(double)a[1]-b[1],z=(double)a[2]-b[2];return Math.sqrt(x*x+y*y+z*z);}
    private static ArrayList<float[]> copy(List<float[]> points){ArrayList<float[]> out=new ArrayList<>();for(float[] p:points)out.add(p.clone());return out;}
    private static ArrayList<float[]> resample(List<float[]> points,int count){
        double[] lengths=new double[points.size()];for(int i=1;i<points.size();i++)lengths[i]=lengths[i-1]+distance(points.get(i-1),points.get(i));
        double total=lengths[lengths.length-1];int segment=1;ArrayList<float[]> result=new ArrayList<>();
        for(int i=0;i<count;i++){if(total<1e-9){result.add(points.get(0).clone());continue;}double at=total*i/(count-1);while(segment<points.size()-1&&lengths[segment]<at)segment++;double length=lengths[segment]-lengths[segment-1];result.add(lerp(points.get(segment-1),points.get(segment),length>1e-9?(at-lengths[segment-1])/length:0));}return result;
    }
    static ArrayList<float[]> smooth(List<float[]> points,float strength,int maxPoints){
        if(points.size()<3||strength<=0)return copy(points);
        boolean closed=distance(points.get(0),points.get(points.size()-1))<1e-6;double amount=Math3.clamp(strength,0,1),length=0;
        for(int i=1;i<points.size();i++)length+=distance(points.get(i-1),points.get(i));if(length<1e-8)return copy(points);
        int count=Math.max(3,Math.min(Math.min(384,maxPoints),Math.max(points.size(),Math.min(128,(int)Math.ceil(length/.015)+1))));
        ArrayList<float[]> rounded=copy(points);
        for(int pass=0;pass<3;pass++){
            ArrayList<float[]> next=new ArrayList<>();if(!closed)next.add(rounded.get(0).clone());
            for(int i=0;i<rounded.size()-1;i++){next.add(lerp(rounded.get(i),rounded.get(i+1),.25));next.add(lerp(rounded.get(i),rounded.get(i+1),.75));}
            next.add((closed?next.get(0):rounded.get(rounded.size()-1)).clone());rounded=next;
        }
        ArrayList<float[]> original=resample(points,count),soft=resample(rounded,count),result=new ArrayList<>();for(int i=0;i<count;i++)result.add(lerp(original.get(i),soft.get(i),amount));
        if(closed)result.set(count-1,result.get(0).clone());else{result.set(0,points.get(0).clone());result.set(count-1,points.get(points.size()-1).clone());}return result;
    }
    static ArrayList<float[]> curve(float[] delta,float[] normal,float bend,boolean sCurve,int count,float firstPressure,float lastPressure){
        ArrayList<float[]> result=new ArrayList<>();float length=Math3.length(delta);if(length<1e-8f){result.add(new float[]{0,0,0,firstPressure});return result;}
        float[] cross=Math3.cross(normal,delta);float n=Math3.length(cross);float[] offset=n>1e-8f?Math3.mul(cross,length*Math3.clamp(bend,-1,1)/n):new float[3];count=Math.max(3,Math.min(65,count));
        for(int i=0;i<count;i++){double t=i/(double)(count-1),u=1-t;float[] p=new float[4];for(int j=0;j<3;j++)p[j]=(float)(delta[j]*t+(sCurve?3*u*t*(u-t):2*u*t)*offset[j]);p[3]=(float)(firstPressure+(lastPressure-firstPressure)*t);result.add(p);}return result;
    }
}

File: web/src/stroke-processing.js
// Authoring operations are baked into the existing editable stroke-point format.
// All interpolation uses convex combinations: no overshoot outside the input hull.
const lerp=(a,b,t)=>a.map((v,i)=>v+(b[i]-v)*t);
const distance=(a,b)=>Math.hypot(a[0]-b[0],a[1]-b[1],a[2]-b[2]);
function resample(points,count){
  const lengths=[0];for(let i=1;i<points.length;i++)lengths.push(lengths[i-1]+distance(points[i-1],points[i]));
  const total=lengths.at(-1);if(total<1e-9)return Array.from({length:count},()=>[...points[0]]);
  let segment=1;return Array.from({length:count},(_,i)=>{const at=total*i/(count-1);while(segment<points.length-1&&lengths[segment]<at)segment++;const length=lengths[segment]-lengths[segment-1];return lerp(points[segment-1],points[segment],length>1e-9?(at-lengths[segment-1])/length:0);});
}
export function smoothPoints(points,strength=.5,maxPoints=384){
  if(points.length<3||strength<=0)return points.map(p=>[...p]);
  const closed=distance(points[0],points.at(-1))<1e-6,amount=Math.max(0,Math.min(1,strength));
  const length=points.slice(1).reduce((sum,p,i)=>sum+distance(p,points[i]),0);
  if(length<1e-8)return points.map(p=>[...p]);
  const count=Math.max(3,Math.min(384,maxPoints,Math.max(points.length,Math.min(128,Math.ceil(length/.015)+1))));
  let rounded=points.map(p=>[...p]);
  for(let pass=0;pass<3;pass++){
    const next=closed?[]:[[...rounded[0]]];
    for(let i=0;i<rounded.length-1;i++){next.push(lerp(rounded[i],rounded[i+1],.25),lerp(rounded[i],rounded[i+1],.75));}
    next.push([...(closed?next[0]:rounded.at(-1))]);rounded=next;
  }
  const original=resample(points,count),soft=resample(rounded,count),result=original.map((p,i)=>lerp(p,soft[i],amount));
  if(closed)result[result.length-1]=[...result[0]];else{result[0]=[...points[0]];result[result.length-1]=[...points.at(-1)];}
  return result;
}
export function curvePoints(delta,normal,bend=.35,kind='arc',count=65,startPressure=1,endPressure=1){
  const length=Math.hypot(...delta);if(length<1e-8)return [[0,0,0,startPressure]];
  const cross=[normal[1]*delta[2]-normal[2]*delta[1],normal[2]*delta[0]-normal[0]*delta[2],normal[0]*delta[1]-normal[1]*delta[0]],n=Math.hypot(...cross);
  const side=cross.map(v=>n>1e-8?v/n:0),offset=side.map(v=>v*length*Math.max(-1,Math.min(1,bend)));
  return Array.from({length:Math.max(3,Math.min(65,count))},(_,i)=>{ // Quadratic arc or cubic S, in the captured plane.
    const t=i/(Math.max(3,Math.min(65,count))-1),u=1-t;
    const p=delta.map((v,j)=>kind==='s'?v*t+3*u*t*(u-t)*offset[j]:v*t+2*u*t*offset[j]);
    return [...p,startPressure+(endPressure-startPressure)*t];
  });
}
